### PR TITLE
Implement support for DWARF version 5.

### PR DESCRIPTION
**Pull request by pcwalton**
DWARF version 5 brings a number of improvements over version 4. Quoting from
the announcement [1]:

> Version 5 incorporates improvements in many areas: better data compression,
> separation of debugging data from executable files, improved description of
> macros and source files, faster searching for symbols, improved debugging
> optimized code, as well as numerous improvements in functionality and
> performance.

On platforms where DWARF version 5 is supported (Linux, primarily), this commit
adds support for it behind a new `-Z dwarf-version=5` flag.

[1]: https://dwarfstd.org/Public_Review.php

r? @michaelwoerister

### CODE REVIEW

**Comment by nikic** [comment]
This should probably be an unstable `-Z dwarf-version` flag, unless you intend this to be insta-stable.

Is the max dwarf version logic really needed? It seems odd that there is no way to force a newer version if a target has this option set. It's reasonable that some platforms default to an older version, but I don't really see a reason to hard forbid using a newer one (for example, if I bring my own toolchain).

**Comment by pcwalton** [comment]
OK, I've switched the `-C` flag to a `-Z` flag, removed the `max_dwarf_version` field from the target info, and moved the documentation over to the unstable book.

**Comment by nnethercote** [comment]
Is this worth a perf run?

File: src/test/assembly/dwarf5.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Makes sure that `-C dwarf-version=5` causes `rustc` to emit DWARF version 5.
```
Comment (nikic) [inline src/test/assembly/dwarf5.rs:1]: `-C` -> `-Z` in this comment.

**Review by nikic — COMMENTED** [review]
Looks reasonable to me.

**Comment by michaelwoerister** [comment]
Thanks, @pcwalton! Looks good to me.

@bors r+

> Is this worth a perf run?

As it currently is, the PR does not actually change the DWARF version being used for anything. It just adds the `-Z` flag for explicitly setting it. So, we shouldn't see a perf difference. But once this is merged it would be interesting to see how switching to v5 affects performance.

**Comment by pcwalton** [comment]
Doesn't look related to this PR.

**Comment by nnethercote** [comment]
That failure is the one that caused the tree closure. Should be resolved now.

@bors retry

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang-ci/rust/runs/7108528451?check_suite_focus=true)

@bors r-

**Comment by pcwalton** [comment]
@bors retry

Looks like that failing test got reworked to not depend on -Z flags.

**Comment by pcwalton** [comment]
Could anyone retry this? I believe the test is fixed now.

**Comment by ehuss** [comment]
@bors r=michaelwoerister

retry does not work when it has been unaccepted.

File: compiler/rustc_codegen_llvm/src/debuginfo/mod.rs [hunk 2]
```diff
@@ -103,14 +103,14 @@ impl<'ll, 'tcx> CodegenUnitDebugContext<'ll, 'tcx> {
...
-                )
-            }
+            let dwarf_version =
+                sess.opts.debugging_opts.dwarf_version.unwrap_or(sess.target.default_dwarf_version);
+            llvm::LLVMRustAddModuleFlag(
+                self.llmod,
+                llvm::LLVMModFlagBehavior::Warning,
+                "Dwarf Version\0".as_ptr().cast(),
+                dwarf_version,
+            );
```
Comment (eddyb) [inline compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:113]: Could this have caused #99143? Looks like before this PR, we only ever set `Dwarf Version` on targets with `Some` in their `dwarf_version`, which based on the diff in this PR, wasn't the case for e.g. Windows (at least, there's probably more targets not covered by the `${OS}_base.rs` files).

Comment (eddyb) [reply compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:113]: Alright that was confusing but I'm pretty sure (https://github.com/rust-lang/rust/issues/99143#issuecomment-1182315377) that what's going on is even with debuginfo disabled for a binary, the standard library is getting statically linked in *and we ship it w/ debuginfo*.

I'm not sure what the best approach here would be, maybe just relying on the `is_like_msvc` check to effectively switch between `CodeView` and `Dwarf Version` (and never set both)?

**Comment by Wyvern** [comment]
Does macOS support DWARF5 at the moment?

**Pull request merged**